Don't modify the provider store when finding a provider.

bob-beck opened the pull request:
Providers are currently kept in the provider store based upon a sorted STACK_OF providers. They are found with sk_OSSL_PROVIDER_find.

We take a write lock on the store every time we find one, because we sort store->providers before searching, every time.

Meanwhile, the only places we add providers to store->providers, we already have the store lock held for write. This change sorts the provider list upon the add of a provider, rather than sorting it every time you look one up. While yes, sorting a sorted stack_of is basically free, the write lock is potentially costly. We potentially do more sorting work in total by sorting the provider list on each add, but presumably we have many more finds than adds, and this list should still be relatively small.

This allows find (which is basically a read operation) to not potentially modify the store, so it need only take a read lock.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

bob-beck committed:
Don't modify the provider store when finding a provider.

Providers are currently kept in the provider store based upon a
sorted STACK_OF providers. They are found with sk_OSSL_PROVIDER_find.

We take a write lock on the store every time we find one, because
we sort store->providers before searching, every time.

Meanwhile, the only places we add providers to store->providers,
we already have the store lock held for write. This change
keeps the provider list sorted when addng providers, rather than
sorting it on lookup

This allows find (which is basically a read operation) to not
potentially modify the store, so it need only take a read lock.

bob-beck commented:
> Member

nvm, I think it's this, found it I beleive.

bob-beck commented:
> This change makes adding providers quadratic time†. That's a foot gun waiting to happen. I'm confident that this is the reason the sorting happened on the find. In other words, I don't think we should accept the PR as it stands.
> 
> Given that lots of providers being loaded isn't going to be a common use case, I'm not sure what the best path forward is. The sorted stack could be replaced by a hash table which would avoid the write lock on find but this seems like overkill. A linear search of an unsorted stack is possible but hits time complexity problems. Unfortunately, we don't have a really lightweight dictionary map.
> 
> † It could even be worse because qsort(3) might not detect that the array is already sorted and do a full _O(n log n)_ sort for each of the _n_ insertions. If qsort(3) does detect the sorted array, this becomes an insertion sort.

I agree it makes adding providers quadratic time. The correct way to fix that would probably be to change this from using a list as the data structure to store providers to something that is sorted on insert, like a red/black tree, however, this would be a more intrusive change for a follow on. 

What I don't necessarily agree with is that making adding providers, where n is small, (i.e. under 100 providers, let's say) and adding a provider is an infrequent operation  is a win relative to causing all lookups to serialize with a write lock, where I believe finding a provider is something that happens more frequently.

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -418,9 +419,10 @@ OSSL_PROVIDER *ossl_provider_find(OSSL_LIB_CTX *libctx, const char *name,
+        if (!CRYPTO_THREAD_read_lock(store->lock))
+            return NULL;
+        if (!sk_OSSL_PROVIDER_is_sorted(store->providers))
             return NULL;
```
Actually - shouldn't we be releasing the lock here?

bob-beck commented on crypto/provider_core.c:
```diff
@@ -418,9 +419,10 @@ OSSL_PROVIDER *ossl_provider_find(OSSL_LIB_CTX *libctx, const char *name,
-        if (!CRYPTO_THREAD_write_lock(store->lock))
+        if (!CRYPTO_THREAD_read_lock(store->lock))
+            return NULL;
+        if (!sk_OSSL_PROVIDER_is_sorted(store->providers))
```
Yes, I actually tried it as such. will change it.

bob-beck commented on crypto/provider_core.c:
```diff
@@ -320,6 +320,7 @@ void *ossl_provider_store_new(OSSL_LIB_CTX *ctx)
         ossl_provider_store_free(store);
         return NULL;
     }
+    sk_OSSL_PROVIDER_sort(store->providers);
```
Yeah, that's sort of a separate change to stack_of.   I assumed it was this way
and then noticed it was not (when the CI failed on my first attempt - basically
the first add ends up failing because the empty stack is not considered sorted.

bob-beck replied:
https://github.com/openssl/openssl/pull/28509

paulidale commented:
I agree that the number of providers will generally be small, that's not an excuse to leave a major problem for the rarer case where lots of providers are loaded.

There is a simple path forward that don't leave the time bomb.  Take a read lock (which you'll need anyway), check for sorted and if not take a write lock and sort.  The write lock contention will only happen at the start.  I.e. something along the lines of:
```c
        if (!CRYPTO_THREAD_read_lock(store->lock))
            return NULL;
        if (!sk_OSSL_PROVIDER_is_sorted(store->providers)) {
            CRYPTO_THREAD_unlock(store->lock);
            if (!CRYPTO_THREAD_write_lock(store->lock))
                return NULL;
            sk_OSSL_PROVIDER_sort(store->providers);
        }

The latter is simpler than the proposed change and avoids introducing the quadratic time issue completely.
```

paulidale commented:
I've done my suggested change over in #28513.
Performance testing is probably required which I'll leave in @bob-beck's hands.

bob-beck commented:
> Since you're looking at doing this as a performance improvement, do we have some before/after metric here that shows the performance gain? It seems fairly evident, but given @paulidale 's comments it would be good to confirm either some improvements or at least a lack of regression.

It's partly as a perfomance improvement, and also partly as a "don't mutate the object on what should be a read operation" - i.e. a const object should work for this.

And yes, I'll have to concoct some sort of performance test.

bob-beck committed:
Don't clear is_sorted unconditionally on OPENSSL_sk_insert()

If we have a comparison function, and the array was sorted,
check to see if we are inserting in the correct location.
if so do not clear is_sorted.

This allows for element locations found with OPENSSL_sk_find_ex
to be used to insert elements in the correct location and preserve
the sorting order without the need to sort the stack again.

paulidale commented on crypto/provider_core.c:
```diff
@@ -673,6 +679,7 @@ int ossl_provider_add_to_store(OSSL_PROVIDER *prov, OSSL_PROVIDER **actualprov,
         if (sk_OSSL_PROVIDER_push(store->providers, prov) == 0)
             goto err;
         prov->store = store;
+        sk_OSSL_PROVIDER_sort(store->providers);
```
Irrespective, locating the element and then inserting will result in O(n²) time behaviour.

Actually it's still O(n² log n) because the find takes O(log n) time.

I.e. doing this isn't a fix.

paulidale commented:
Doing this using an RCU hash for lookup seems like the best solution.  I don't think we should fiddle around and kludge a partial solution.

beldmit reviewed (approved):
LGTM

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 124 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 155 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 186 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 217 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 248 days ago
